### ISSUE TITLE
Rule idea: no spreading props

### ISSUE THREAD

**Issue by mockdeep** [yannickcr/eslint-plugin-react#1094]
One thing we've noticed tends to make code a little opaque is when we destructure the entire props for a sub-component:

```jsx
function MyComponent(props) {
  return <MySubComponent {...props} />
}
```

It would be great to have an eslint rule to force expanding individual props that should be passed down:

```jsx
function MyComponent(props) {
  return <MySubComponent one_prop={props.one_prop} two_prop={props.two_prop} />
}
```

**Comment by ljharb** [comment]
In other words, forbidding spread props, but just of the entire props object?

This seems pretty useful. I'd actually want to forbid spread props in all cases, so that maybe could be configurable.

**Comment by mockdeep** [comment]
Yeah, a configuration would be good, I think. Sometimes we like to assign local props or use a function to pull them out, which I think we'd like to still be able to do.

```jsx
function MyComponent(props) {
  return <MySubComponent {...subComponentProps(props)} />
}

function subComponentProps(props) {
  return {
    one_prop,
    two_prop: two_prop * 2
  };
}
```

Haven't landed on a clear style for that, though.

**Comment by thoiberg** [comment]
If no one else is actively working on this I wouldn't mind taking a shot. I've run into a few cases where as components have grown using `{...this.props}` has ended up passing in 10 props to a child component when it only needs 3.

**Comment by ljharb** [comment]
@thoiberg you can use https://npmjs.com/prop-types-exact on each of your components to help catch that, but the linter rule is also great :-)

**Comment by sanjayp** [comment]
Just adding that I'd like to see this rule added as well since there are potential performance implications, as noted here: https://flexport.engineering/optimizing-react-rendering-part-1-9634469dca02#63e7.

**Comment by thoiberg** [comment]
@benmonro no sorry, I got a little way's in but work got in the way 😞. Feel free to take over if you have the desire 😄

**Comment by iegik** [comment]
I think that rule was about:

```
<Foo {...{
  bar: 'tar',
  moo: 0,
}} />
```

same as:

```
<Foo bar="tar" moo={0} />
```

**Comment by iegik** [comment]
@mockdeep

```
function MyComponent(props) {
  return <MySubComponent {...subComponentProps(props)} />
}
```
->
```
function MyComponent(props) {
  const { one_prop, two_prop } = subComponentProps(props);

  return <MySubComponent one_prop={one_prop} two_prop={two_prop} />
}
```

**Comment by mockdeep** [comment]
@iegik yeah, that could work if it's too hard to implement my suggestion above.

**Comment by KutnerBitsrc** [comment]
yes this is needed.
A lot of places in our code have {...props} and it makes the code unmaintainable

**Comment by ashbhir** [comment]
I can take this up, will raise a PR soon

**Comment by cwilby** [comment]
Fun side effect, I've inherited a project where this rule became active after it was built.  It contains a Button component that spreads props (assuming in case there's a need to pass native props).

Should I refactor this component to handle all native properties of an HTMLInputElement or ignore this rule?  I can't really justify using button directly given the use case.

**Comment by KutnerUri** [comment]
This is more or less solved by Typescript.
A year ago, when I posted the last comment, I was working on a legacy project with half working Flow types. The _no-jsx-spread rule would still be useful in that project_.

In my current project, I no longer have this problem, as I am passing and combining types for props.

In fact, in a type-safe world, it actually **makes more sense** to spread props to allow the **full range of native props** with **auto complete**.

```tsx
type LoaderButtonProps = ButtonHTMLAttributes<HTMLButtonElement>
  & { customProps... };

export default function LoaderButton(props: LoaderButtonProps) {
    const [ isLoading, setLoading ] = useState(false);
    //(this could be optimised, I know)
    const handleClick = props.onClick && (e: MouseEvent<...>) => {
        setLoading(true);
        props.onClick(e).then(() => setLoading(false));
    }

    if(isLoading) return <Loader/>;

    return <button {...props} onClick={handleClick} />
}
```

**Comment by ljharb** [comment]
This presumes that every component is defined to only allow exact props (which they’re not by default in TS) and that TS types are capable of fully expressing the type for all JS values (it isn’t), and that all components you use are authored in TS (unlikely if you use anything from npm).

**Comment by KutnerUri** [comment]
I agree.
When these conditions are not met the _no-jsx-spread_  rule should apply. (even in entire project)

in TS files, I found this pattern to be useful enough:
```tsx
type SuperProps = {
    rainbow: boolean;
}

type SuperButtonProps = LoaderButtonProps & SuperProps;

function SuperButton(props: SuperButtonProps) {
    const { rainbow, children, ...baseProps } = props;

    return <LoaderButton {...baseProps} >
        {rainbow && <Rainbows/>}
        {children}
    </LoaderButton>;
}
```

**Comment by ljharb** [comment]
I think you should explicitly pass the props you need, only.
